### PR TITLE
Change all event containers to store objects

### PR DESCRIPTION
**Pull request by Mkalo**
Well first I've changed the Event pointers to use smart pointers, the
second step would be to change the registerEvent functions to be
something like TalkActions::registerEvent (being careful to copy or move
depending on the event if it allows duplicates etc.)

But before doing that I would like some opinion on that, like if it does
even have a chance of being merged.

### CODE REVIEW

**Comment by ranisalt** [comment]
I do appreciate it. There's a land of undefined behavior and segmentation fault on naked pointers.

File: src/baseevents.cpp [hunk 1]
```diff
@@ -61,15 +61,19 @@ bool BaseEvents::loadFromXml()
...
+		bool success;
+
 		pugi::xml_attribute scriptAttribute = node.attribute("script");
 		if (scriptAttribute) {
 			std::string scriptFile = "scripts/" + std::string(scriptAttribute.as_string());
-			event->checkScript(basePath, scriptsName, scriptFile) && event->loadScript(basePath + scriptFile);
+			success = event->checkScript(basePath, scriptsName, scriptFile) && event->loadScript(basePath + scriptFile);
 		} else {
-			event->loadFunction(node.attribute("function"));
+			success = event->loadFunction(node.attribute("function"));
```
Comment (ranisalt) [inline src/baseevents.cpp]: Functions are generally being removed, it would be better to wait until they are removed so you won't need this book :)

File: src/creatureevent.cpp [hunk 2]
```diff
@@ -57,17 +57,17 @@ std::string CreatureEvents::getScriptBaseName() const
...
 }
 
-Event* CreatureEvents::getEvent(const std::string& nodeName)
+std::unique_ptr<Event> CreatureEvents::getEvent(const std::string& nodeName)
 {
 	if (strcasecmp(nodeName.c_str(), "event") != 0) {
 		return nullptr;
 	}
-	return new CreatureEvent(&scriptInterface);
+	return std::unique_ptr<Event>(new CreatureEvent(&scriptInterface));
```
Comment (ranisalt) [inline src/creatureevent.cpp]: `return {new ...};` is cleaner

File: src/creatureevent.cpp [hunk 3]
```diff
@@ -57,17 +57,17 @@ std::string CreatureEvents::getScriptBaseName() const
...
 }
 
-bool CreatureEvents::registerEvent(Event* event, const pugi::xml_node&)
+bool CreatureEvents::registerEvent(std::unique_ptr<Event>&& event, const pugi::xml_node&)
 {
-	CreatureEvent* creatureEvent = static_cast<CreatureEvent*>(event); //event is guaranteed to be a CreatureEvent
+	CreatureEvent* creatureEvent = static_cast<CreatureEvent*>(event.release()); //event is guaranteed to be a CreatureEvent
```
Comment (ranisalt) [inline src/creatureevent.cpp]: Why not unique ptr here too?

Comment (Mkalo) [inline src/creatureevent.cpp]: Yea, I will change all those pointers to unique ptr when I start changing the containers. I let it like that for now.

File: src/actions.cpp [hunk 4]
```diff
@@ -76,17 +76,17 @@ std::string Actions::getScriptBaseName() const
...
 	}
-	return new Action(&scriptInterface);
+	return std::unique_ptr<Event>(new Action(&scriptInterface));
 }
 
-bool Actions::registerEvent(Event* event, const pugi::xml_node& node)
+bool Actions::registerEvent(std::unique_ptr<Event>&& event, const pugi::xml_node& node)
 {
-	Action* action = static_cast<Action*>(event); //event is guaranteed to be an Action
+	Action* action = static_cast<Action*>(event.release()); //event is guaranteed to be an Action
```
Comment (djarek) [inline src/actions.cpp]: Violates http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#a-namerf-consumeaf18-for-consume-parameters-pass-by-x-and-stdmove-the-parameter ?

File: src/globalevent.cpp [hunk 5]
```diff
@@ -60,17 +60,17 @@ void GlobalEvents::clear()
...
 }
 
-Event* GlobalEvents::getEvent(const std::string& nodeName)
+std::unique_ptr<Event> GlobalEvents::getEvent(const std::string& nodeName)
 {
 	if (strcasecmp(nodeName.c_str(), "globalevent") != 0) {
 		return nullptr;
 	}
-	return new GlobalEvent(&scriptInterface);
+	return std::unique_ptr<Event>(new GlobalEvent(&scriptInterface));
```
Comment (djarek) [inline src/globalevent.cpp]: ```return {new ...};```

File: src/spells.cpp [hunk 6]
```diff
@@ -122,31 +122,33 @@ std::string Spells::getScriptBaseName() const
...
 	return nullptr;
 }
 
-bool Spells::registerEvent(Event* event, const pugi::xml_node&)
+bool Spells::registerEvent(std::unique_ptr<Event>&& event, const pugi::xml_node&)
 {
-	InstantSpell* instant = dynamic_cast<InstantSpell*>(event);
+	InstantSpell* instant = dynamic_cast<InstantSpell*>(event.get());
 	if (instant) {
+		event.release();
```
Comment (djarek) [inline src/spells.cpp]: Don't do this because it looks like a leak at first glance.

File: src/creatureevent.cpp (hunk 2 above)
Comment (djarek) [inline src/creatureevent.cpp]: Conversion constructor is explicit.

**Comment by DSpeichert** [comment]
@djarek Please re-review if you have time.

File: src/weapons.h [hunk 7]
```diff
@@ -62,7 +62,7 @@ class Weapons final : public BaseEvents
 class Weapon : public Event
 {
 	public:
-		explicit Weapon(LuaScriptInterface* interface) : Event(interface) {}
```
Comment (djarek) [inline src/weapons.h]: Why did you remove the `explicit` specifier?

Comment (Mkalo) [reply src/weapons.h]: Not really sure, I'll re add them

File: src/actions.cpp [hunk 8]
```diff
@@ -108,14 +99,14 @@ bool Actions::registerEvent(Event* event, const pugi::xml_node& node)
 		uint16_t iterId = fromId;
 		uint16_t toId = pugi::cast<uint16_t>(toIdAttribute.value());
 
-		auto result = useItemMap.emplace(iterId, action);
+		auto result = useItemMap.emplace(iterId, *action);
```
Comment (djarek) [inline src/actions.cpp]: Why aren't you moving the action, like in line 86?

Comment (Mkalo) [reply src/actions.cpp]: Because in this part, its the fromid and toid code, I need to make a copy of the action for everytime it's inserted. Not ideal but I don't know how to do it better with a map.
